internal/versionsapi/cli/rm.go

```diff
@@ -0,0 +1,674 @@
+	stream, err := cmd.Flags().GetString("stream")
+	if err != nil {
+		return nil, err
+	}
+	version, err := cmd.Flags().GetString("version")
+	if err != nil {
+		return nil, err
+	}
+	versionPath, err := cmd.Flags().GetString("version-path")
+	if err != nil {
+		return nil, err
+	}
+	all, err := cmd.Flags().GetBool("all")
+	if err != nil {
+		return nil, err
+	}
+	dryrun, err := cmd.Flags().GetBool("dryrun")
+	if err != nil {
+		return nil, err
+	}
+	region, err := cmd.Flags().GetString("region")
+	if err != nil {
+		return nil, err
+	}
+	bucket, err := cmd.Flags().GetString("bucket")
+	if err != nil {
+		return nil, err
+	}
+	distributionID, err := cmd.Flags().GetString("distribution-id")
+	if err != nil {
+		return nil, err
+	}
+	gcpProject, err := cmd.Flags().GetString("gcp-project")
+	if err != nil {
+		return nil, err
+	}
+	azSubscription, err := cmd.Flags().GetString("az-subscription")
+	if err != nil {
+		return nil, err
+	}
+	azLocation, err := cmd.Flags().GetString("az-location")
+	if err != nil {
+		return nil, err
+	}
+	azResourceGroup, err := cmd.Flags().GetString("az-resource-group")
+	if err != nil {
+		return nil, err
+	}
+	debug, err := cmd.Flags().GetBool("debug")
+	if err != nil {
+		return nil, err
+	}
+	logLevel := zapcore.InfoLevel
+	if debug {
+		logLevel = zapcore.DebugLevel
+	}
+
+	return &rmFlags{
+		ref:             ref,
+		stream:          stream,
+		version:         version,
+		versionPath:     versionPath,
+		all:             all,
+		dryrun:          dryrun,
+		region:          region,
+		bucket:          bucket,
+		distributionID:  distributionID,
+		gcpProject:      gcpProject,
+		azSubscription:  azSubscription,
+		azLocation:      azLocation,
+		azResourceGroup: azResourceGroup,
+		logLevel:        logLevel,
+	}, nil
+}
+
+type awsClient struct {
+	ec2 ec2API
+}
+
+// newAWSClient creates a new awsClient.
+// Requires IAM permission 'ec2:DeregisterImage'.
+func newAWSClient(ctx context.Context, region string) (*awsClient, error) {
+	return &awsClient{}, nil
+}
+
+type ec2API interface {
+	DeregisterImage(ctx context.Context, params *ec2.DeregisterImageInput, optFns ...func(*ec2.Options),
+	) (*ec2.DeregisterImageOutput, error)
+}
+
+func (a *awsClient) deleteImage(ctx context.Context, ami string, region string, dryrun bool, log *logger.Logger) error {
+	cfg, err := awsconfig.LoadDefaultConfig(ctx, awsconfig.WithRegion(region))
+	if err != nil {
+		return err
+	}
+	a.ec2 = ec2.NewFromConfig(cfg)
+
+	req := ec2.DeregisterImageInput{
+		ImageId: &ami,
+		DryRun:  &dryrun,
+	}
+
+	log.Debugf("Deregistering image %s", ami)
+	_, err = a.ec2.DeregisterImage(ctx, &req)
+	var apiErr smithy.APIError
+	if errors.As(err, &apiErr) &&
+		(apiErr.ErrorCode() == "InvalidAMIID.NotFound" ||
+			apiErr.ErrorCode() == "InvalidAMIID.Unavailable") {
+		log.Warnf("AWS image %s not found", ami)
+	} else if err != nil {
+		return fmt.Errorf("deregistering image %s: %w", ami, err)
+	}
+
+	return nil
+}
```

Comment:
This leaves a dangling snapshot.

Deletion should:

- `get` AMI
- parse corresponding snapshot
- deregister ami
- delete snapshot